### ISSUE TITLE
Eval runs are not reliably deterministic: one of two re-run cells diverged, and three places assume otherwise

### ISSUE THREAD

**Issue by chrisshaiman** [chrisshaiman/lamware#381]
Re-running two cells produced **one byte-identical result and one that diverged**. The harness treats determinism as load-bearing in at least three places, and one of them gates a refusal.

## Measured

Same corpus, same arm, same deployed code, re-run within hours:

| sample | calls | out_tok | schema | parse_note | identical? |
|---|---|---|---|---|---|
| emotet `d7efad30d0a5` | 12 | 739 | 6/9 | yes | **byte-identical** |
| trickbot `bf96e2c4f383` | 10 | 667 | 6/9 | yes | **narrative differs** |

The trickbot divergence, at an identical output-token count:

```
original: "capabilities": [ "String manipulation (MFC CStringT)", "Memory management/
rerun:    "capabilities": [ "Memory manipulation (memcpy)",     "Object lifecycle management
```

## Why this matters more than two cells

`lamware_eval/__main__.py:59` refuses a consensus axis on the grounds that:

> Runs are **deterministic** (#292), so qwen@10's trajectory is a literal PREFIX of qwen@15's on the same sample

`consensus.py:26` reasons from the same premise. And the whole justification for `n=1` per cell — that a single run *is* the model's output for a config, so there is nothing to average — rests on it.

If determinism is conditional, then:

- **n=1 is no longer self-justifying.** Cells may need repetition, which multiplies an already 4-hour sweep.
- The depth-prefix argument behind the consensus refusal is weakened. The refusal may still be right, but not for the stated reason.
- The earlier raccoonstealer re-baseline result I reported as "byte-identical, determinism holds exactly" was **narrower than I claimed** — it was one sample that happened to match, not a demonstrated property.

## Likely mechanism (inferred, not measured)

The server runs with:

```
--seed 42 --temp 0.6 --top-p 0.95 --top-k 20
```

That is **stochastic sampling with a single seed set once at process start**. The sequence is reproducible from a fresh start, but the RNG advances across requests — so the same request produces the same tokens only if it occupies the same position in the server's stream. Request count and ordering differ between a 29-cell sweep and a 2-cell re-run, and `llama-cpp` has restarted 4 times (`NRestarts=4`, including a `status=137` SIGKILL mid-sweep).

That would explain both observations: identical output when the stream happens to line up, divergence when it does not.

**This is a hypothesis.** It should be tested directly before anything is built on it — two identical requests back to back on a freshly started server, then the same pair with an unrelated request injected between them.

## Suggested

1. **Test the mechanism** as above. Cheap and decisive.
2. **Stop asserting determinism in comments and refusal messages** until it is characterised. `__main__.py:59` currently tells the operator something we have now seen to be false.
3. Decide deliberately: either make runs actually reproducible (`--temp 0` for eval arms, or per-request seeding if the transport ever supports it — note #292 found `seed` is ignored on `/v1/messages`), or accept non-determinism and require n>1 for any claim that rests on a single cell.

Option one is probably right for a benchmark: `--temp 0.6` is a *serving* choice that leaked into a *measurement* context.

## Separately: the truncation reproduces

Both cells failed the same way on both runs — same call count, same output-token count, same missing schema fields. So the parse failures (#380) are a **reproducible behaviour on these inputs**, not a flake and not the OOM restart. That remains unexplained and is tracked separately.

**Comment by chrisshaiman** [comment]
## The RNG-position hypothesis is falsified

Tested on the idle server, hashing the full response content (thinking included), 1500 max_tokens:

| test | result |
|---|---|
| **A** same request, back to back | `b586c3c950be` = `b586c3c950be` — identical |
| **B** same request, unrelated request injected between | `b586c3c950be` = `b586c3c950be` — **identical** |
| **C** temperature=0, back to back | `b4201670ae25` ×2 — identical |
| **D** temperature=0, interleaved | `b4201670ae25` ×2 — identical |

Injecting an unrelated request between two identical ones changed nothing. **Request position in the server's RNG stream does not affect output.** The proposed mechanism in the issue body is wrong and should not be acted on.

Two useful by-products:

- **`--temp 0.6` is not the problem.** Sampling is already reproducible at 738 output tokens without touching temperature, so the "switch eval arms to temp 0" fix would have changed nothing. Good that this was tested before the config change.
- **Per-request `temperature` IS honoured** (A/B and C/D produce different hashes). That is worth knowing on its own, because #292 established the transport silently *ignores* per-request `seed` — so the two parameters behave differently and the difference is not documented anywhere.

## Revised hypothesis: context size / KV cache state

What the falsified test did not reproduce is the eval's shape. It sent a single-turn request with a small prompt. The diverging cell was a **multi-turn agentic loop** with a large accumulated context.

The pattern in the two re-run cells supports this:

| sample | functions | imports | re-run |
|---|---|---|---|
| emotet `d7efad30d0a5` | 3,502 | 126 | **byte-identical** |
| trickbot `bf96e2c4f383` | 3,603 | **568** | **diverged** |

The one that diverged is the one with the largest prompt — 568 imports in the initial message versus 126.

A plausible mechanism, **stated as untested**: llama.cpp prefills in chunks, and chunk boundaries depend on how much of the prefix is already cached. Floating-point addition is not associative, so a different accumulation order yields marginally different logits, and at a near-tie between two tokens that flips the sample. The server also restarted 4 times (`NRestarts=4`, including a `status=137` SIGKILL mid-sweep), which resets cache state entirely between the original run and the re-run.

That predicts: **determinism holds for small prompts and degrades as context grows** — which is exactly backwards from what a benchmark wants, since the large-context samples are the interesting ones.

## Next test

Repeat A and B with a prompt sized to the eval's actual context (tens of thousands of tokens) rather than a few hundred. If large prompts diverge where small ones do not, the cache/chunking explanation stands and `temp 0` is confirmed irrelevant.

Until then the practical guidance is unchanged and the issue stays open: **do not assert determinism in comments or refusal messages** (`__main__.py:59`, `consensus.py:26`), because it demonstrably does not hold for at least one real eval cell — whatever the cause.

**Comment by chrisshaiman** [comment]
Two experiments. The headline: **nondeterminism does not reproduce under controlled conditions**, and one of the premises this issue rested on was wrong.

## 1. The seed arms are not evidence of nondeterminism

raccoonstealer has five `qwen@10` variants whose claim counts differ (15 / 16 / 9 / 9 / 9), which looked like five runs of one config diverging. Timestamps say otherwise:

```
qwen_10_s42.variance   2026-07-27 21:03:59  ┐
qwen_10_s1337          2026-07-27 21:17:49  ├─ BYTE-IDENTICAL
qwen_10_s8675309       2026-07-27 21:31:40  ┘
qwen_10_s42            2026-07-31 00:40:29   +4 days
qwen_10                2026-08-09 07:55:39   +9 days
```

The three that ran consecutively on one evening produced **identical tool-call sequences and identical narratives** (`narr_sha=bbb09d39e6`, `out=772`, 10 calls) despite three different seeds. The two that differ ran days apart across many merges, so they are uncontrolled.

That is a clean confirmation of #292 from a new angle — **the seed does nothing** — and evidence *for* determinism within a session, not against it.

## 2. Live A/B at eval-scale context: deterministic

Same request against LiteLLM `/v1/messages`, the production transport, at a 9,035-token prompt:

| test | result |
|---|---|
| A — consecutive | **identical** (`3387b46fa1ea`) |
| B — unrelated request interleaved | **identical** (`3387b46fa1ea`) |
| C — small-prompt control | **identical** (`ae020034aaa2`) |

KV-cache reuse is plainly visible: the first request took **669.8s**, every later one **68.6s**. And the server log shows the mechanism by name:

```
slot get_availabl: selected slot by LCP similarity, sim_best = 0.863 (> 0.100 thold)
```

llama.cpp picks a slot by longest-common-prefix similarity to prior traffic and reuses that KV state — so a preceding request *can* change a later one's numerics. It just didn't here, even with an unrelated request deliberately interleaved.

### What this test does not show

- Generation stopped at `out=600`, exactly `max_tokens`. It compares 600 tokens of thinking, not a complete answer. A real comparison (1,870 chars hashed, not the empty-string trap), but truncated.
- 9k tokens is below the eval's larger prompts — 33–36k on the seeded raccoonstealer arms. **The context-size hypothesis is narrowed, not falsified.**
- LiteLLM reported `input_tokens` as 4 and 17 for a 9k-token prompt. Its usage accounting is wrong; irrelevant to the comparison, but do not trust those numbers elsewhere.

## 3. The remaining candidate mechanism

`llama-cpp` was SIGKILLed mid-sweep and restarted:

```
2026-08-11T00:52:39  Main process exited, code=exited, status=137
2026-08-11T00:52:49  Started llama-cpp.service   (restart counter: 4)
```

The sweep ran straight through it — `icedid_0d2c47c8124f` finished 00:47:53, `icedid_12d631f48dd5` 00:54:17. A restart drops all KV state, so cells either side of it are not comparable.

That fits every observation: identical within a session, divergent across days and restarts.

I could not locate the two cells this issue originally cited (`d7efad30d0a5`, `bf96e2c4f383`) anywhere on disk, so I cannot re-check whether they straddled a restart. Stating that rather than assuming.

## Proposed disposition

The three places that assert determinism are still wrong, but for a narrower reason than "runs are nondeterministic":

- `lamware_eval/__main__.py:59` and `consensus.py:26` tell the operator runs are deterministic **and therefore** a qwen@10 trajectory is a literal prefix of qwen@15. Determinism holding *within a session* does not license comparing runs from different sessions, which is what a sweep does.
- The n=1 justification has the same gap.

Suggest rewording those to "deterministic within a single server session; runs separated by a restart are not comparable", and recording the server's `ActiveEnterTimestamp` in each cell's result so a sweep can tell which cells are comparable at all. That is a smaller and more defensible claim than either "deterministic" or "nondeterministic".

Still open before closing: repeat the A/B at 36k tokens, matching the largest real prompts.
